File: problems/chapter16/nicetak/arc085_c.jl
mutable struct Edge
    rev::Int
    from::Int
    to::Int
    cap::Int
end

function push!(G::Vector{Vector{Edge}}, from, to, cap)
    fromrev = length(G[from]) + 1
    torev = length(G[to]) + 1
    Base.push!(G[from], Edge(torev, from, to, cap))
    Base.push!(G[to], Edge(fromrev, to, from, 0))
end

function run_flow!(G::Vector{Vector{Edge}}, e::Edge, f)
    e.cap -= f
    G[e.to][e.rev].cap += f
    return G
end

function fodfs(G, v, t, f, seens)
    v == t && return f

    seens[v] = true
    for e ∈ G[v]
        if seens[e.to] || e.cap == 0
            # Skip
        else
            flow = fodfs(G, e.to, t, min(f, e.cap), seens)
            if flow == 0
                # Skip
            else
                run_flow!(G, e, flow)
                return flow
            end
        end
    end

    return 0
end

function FordFulkerson(G, s, t)
    INF = 10^9
    res = 0

    while true
        flow = fodfs(G, s, t, INF, falses(length(G)))
        if flow == 0
            return res
        else
            res += flow
        end
    end

    return 0
end


function main()
    N = parse(Int, readline())
    as = parse.(Int, split(readline()))
    INF = 10^9

    offset = 0
    G = [Vector{Edge}() for _ = 1:N+2]
    for (i, a) = enumerate(as)
        if a ≥ 0
            push!(G, N+1, i, 0)
            push!(G, i, N+2, a)
            offset += a
        else
            push!(G, N+1, i, -a)
            push!(G, i, N+2, 0)
        end
    end

    for i = 1:N
        for j = i+1:N
            if j % i == 0
                push!(G, i, j, INF)
            end
        end
    end

    res = FordFulkerson(G, N+1, N+2)
    println(offset - res)
end

main()

File: problems/chapter16/tokuma/arc085_c.cpp
// https://atcoder.jp/contests/arc085/tasks/arc085_c
#include <iostream>
#include <vector>

using namespace std;

// グラフを表す構造体
struct Graph
{
    // 辺を表す構造体
    // rev: 逆辺 (to, from) が G[to] の中で何番目の要素か
    // cap: 辺 (from, to) の容量
    struct Edge
    {
        int rev, from, to;
        long long cap;
        Edge(int r, int f, int t, long long c) : rev(r), from(f), to(t), cap(c) {}
    };

    // 隣接リスト
    vector<vector<Edge>> list;

    // N: 頂点数
    Graph(int N = 0) : list(N) {}

    // グラフの頂点数取得
    size_t size()
    {
        return list.size();
    }

    // Graph インスタンスを G として，
    // G.list[v] を G[v] と書けるようにしておく
    vector<Edge> &operator[](int i)
    {
        return list[i];
    }

    // 辺 e = (u, v) の逆辺 (v, u) を取得する
    Edge &redge(const Edge &e)
    {
        return list[e.to][e.rev];
    }

    // 辺 e = (u, v) に流量 f のフローを流す
    // e = (u, v) の流量が f だけ減少する
    // このとき逆辺 (v, u) の流量を増やす
    void run_flow(Edge &e, long long f)
    {
        e.cap -= f;
        redge(e).cap += f;
    }

    // 頂点 from から頂点 to へ容量 cap の辺を張る
    // このとき to から from へも容量 0 の辺を張っておく
    void addedge(int from, int to, long long cap)
    {
        int fromrev = (int)list[from].size();
        int torev = (int)list[to].size();
        list[from].push_back(Edge(torev, from, to, cap));
        list[to].push_back(Edge(fromrev, to, from, 0LL));
    }
};

struct FordFulkerson
{
    static const long long INF = 1LL << 55; // 無限大を表す値を適切に
    vector<int> seen;

    FordFulkerson() {}

    // 残余グラフ上で s-t パスを見つける (深さ優先探索)
    // 返り値は s-t パス上の容量の最小値 (見つからなかったら 0)
    // f: s から v へ到達した過程の各辺の容量の最小値
    long long fodfs(Graph &G, int v, int t, long long f)
    {
        // 終端 t に到達したらリターン
        if (v == t)
            return f;

        // 深さ優先探索
        seen[v] = true;
        for (auto &e : G[v])
        {
            if (seen[e.to])
                continue;

            // 容量 0 の辺は実際には存在しない
            if (e.cap == 0LL)
                continue;

            // s-t パスを探す
            // 見つかったら flow はパス上の最小容量
            // 見つからなかったら f = 0
            long long flow = fodfs(G, e.to, t, min(f, e.cap));

            // s-t パスが見つからなかったら次辺を試す
            if (flow == 0LL)
                continue;

            // 辺 e に容量 flow のフローを流す
            G.run_flow(e, flow);

            // s-t パスを見つけたらパス上最小容量を返す
            return flow;
        }

        // s-t パスが見つからなかったことを示す
        return 0LL;
    }

    // グラフ G の s-t 間の最大流量を求める
    // ただしリターン時に G は残余グラフになる
    long long solve(Graph &G, int s, int t)
    {
        long long res = 0;

        // 残余グラフに s-t パスがなくなるまで反復
        while (true)
        {
            seen.assign((int)G.size(), 0);
            int flow = fodfs(G, s, t, INF);

            // s-t パスが見つからなかったら終了
            if (flow == 0LL)
                return res;

            // 答えを加算
            res += flow;
        }

        // no reach
        return 0LL;
    }
};
const long long INF = 1LL << 55;
int main()
{
    int N;
    cin >> N;

    Graph G(N + 2);
    vector<long long> a(N);

    for (int i = 0; i < N; i++)
    {
        cin >> a[i];
    }

    // スーパーノード
    int s = N;
    int t = N + 1;

    long long max_val = 0;

    // グラフの作成
    // 最小コストを計算する形に帰着する
    for (int i = 0; i < N; ++i)
    {
        // aの値で場合わけ
        if (0 <= a[i])
        {

            // 割る操作をしないときのコストが0
            G.addedge(s, i, 0);

            // 割る操作をするときのコストがa[i]
            G.addedge(i, t, a[i]);
            max_val += a[i];
        }
        else
        {
            // 割る操作をしないときのコストが-a[i]
            G.addedge(s, i, -a[i]);

            // 割る操作をするときのコストが0
            G.addedge(i, t, 0);
        }
    }

    //　制約条件
    for (int i = 0; i < N; ++i)
    {
        for (int j = i + 1; j < N; ++j)
        {
            // 倍数の場合にはどちらかのみ押すということができないようにする。
            if (((j + 1) % (i + 1)) == 0)
            {
                G.addedge(i, j, INF);
            }
        }
    }

    // sからtにかけて
    FordFulkerson ff;
    cout << max_val - ff.solve(G, s, t) << endl;
}
